### PR TITLE
Phase 5: Transolver++ Without torch.compile — Full Training Budget (8 parallel)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Run the Transolver++ features (Gumbel-Softmax + per-point adaptive temperature) with torch.compile DISABLED. Edward's T3 attempt showed compile takes 114min of 180min — disabling it gives the full training budget.

tanjiro showed Gumbel+AdaTemp (without memory optimization) was the best config at val/loss 0.3957. This PR validates that config with more epochs by disabling compile.

## Instructions

Use tanjiro's winning config (gumbel_scale=0.1 + ada_temp, NO tpp_no_fx_proj) with --compile_mode disable.

| GPU | Config | Notes |
|-----|--------|-------|
| 0-3 | Gumbel+AdaTemp, no compile | seeds 42-45 |
| 4-5 | Gumbel only, no compile | seeds 42-43 |
| 6-7 | Baseline, no compile | seeds 42-43 |

## Baseline
val/loss 0.401, p_in 12.95, p_oodc 8.40, p_tan 33.8, p_re 24.7

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT thorfinn:

## Results

### Summary

**Disabling torch.compile is catastrophic.** All no-compile runs completed only ~68-74 epochs vs ~160 with compile in the same 180-min window. This halved the training budget and every config (including baseline) performed much worse than the compiled baseline.

The Transolver++ features (Gumbel-Softmax + AdaTemp) cannot be fairly evaluated in this context since all runs are severely under-trained.

### Detailed Metrics

| GPU | Config | Epochs | val/loss | p_in | p_oodc | p_tan | p_re |
|-----|--------|--------|----------|------|--------|-------|------|
| — | **Compiled Baseline (8-seed mean)** | **160** | **0.400** | **13.03** | **8.23** | **33.54** | **24.56** |
| 0 | Gumbel+AdaTemp, nocompile, s42 | 68 | 0.533 | 21.98 | 14.50 | 38.18 | 28.67 |
| 1 | Gumbel+AdaTemp, nocompile, s43 | 68 | 0.584 | 27.22 | 14.63 | 44.24 | 28.57 |
| 2 | Gumbel+AdaTemp, nocompile, s44 | 68 | 0.539 | 22.65 | 14.15 | 37.99 | 28.85 |
| 3 | Gumbel+AdaTemp, nocompile, s45 | 68 | 0.565 | 25.55 | 14.54 | 40.33 | 29.38 |
| 4 | Gumbel only, nocompile, s42 | 69 | 0.581 | 25.92 | 17.12 | 40.12 | 30.59 |
| 5 | Gumbel only, nocompile, s43 | 69 | 0.542 | 21.72 | 15.26 | 37.93 | 28.89 |
| 6 | Baseline, nocompile, s42 | 74 | 0.529 | 23.85 | 13.84 | 40.14 | 27.96 |
| 7 | Baseline, nocompile, s43 | 74 | 0.530 | 21.43 | 14.74 | 37.73 | 29.07 |

### Per-epoch timing

- **With compile**: ~75s/epoch (after ~10 min compilation warmup) → 160 epochs in 180 min
- **Without compile**: ~148-160s/epoch (no warmup, but 2x slower) → 68-74 epochs in 180 min

torch.compile provides a **~2x speedup per epoch** after the initial compilation pass. The compilation overhead is only ~10-15 minutes (not 114 min as the PR hypothesis suggested), so the compiled model gets significantly more training.

### Gumbel+AdaTemp vs Baseline (within no-compile runs)

Within the no-compile runs, comparing apples-to-apples:

| Config | Mean val/loss | Mean p_in |
|--------|--------------|-----------|
| Gumbel+AdaTemp (4 seeds) | 0.555 ± 0.024 | 24.35 ± 2.33 |
| Gumbel only (2 seeds) | 0.561 ± 0.028 | 23.82 ± 2.97 |
| Baseline nocompile (2 seeds) | 0.530 ± 0.001 | 22.64 ± 1.72 |

The baseline without Gumbel/AdaTemp actually performed slightly better than the Transolver++ features in the no-compile setting. This may be because the Gumbel/AdaTemp features add parameters that need more training to converge, and 68 epochs is insufficient.

### VRAM: 45.7 GB for all configs (no compile → no graph optimization → higher memory)

### Code changes

1. Added Gumbel-Softmax noise (`--tpp_gumbel`, `--tpp_gumbel_scale`) to slice routing during training
2. Added per-point adaptive temperature (`--tpp_ada_temp`) with zero-initialized projection
3. Added `--tpp_gumbel_anneal` for annealing Gumbel scale during training
4. Added `compile_mode=disable` option to skip torch.compile
5. Added `random.seed()` and `numpy.random.seed()` for full reproducibility
6. Added `node_subsample_frac` (from PR #1901 implementation on this branch)

### Commands used

```bash
cd cfd_tandemfoil && CUDA_VISIBLE_DEVICES=X python train.py \
  --agent thorfinn --wandb_name "thorfinn/<variant>" \
  --wandb_group phase5/transolver-pp-nocompile \
  --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp \
  --n_layers 3 --slice_num 96 --tandem_ramp \
  --domain_layernorm --domain_velhead --ema_decay 0.999 \
  --weight_decay 5e-5 --cosine_T_max 180 --disable_pcgrad \
  --pressure_first --pressure_deep --compile_mode disable \
  [--tpp_gumbel --tpp_gumbel_scale 0.1] [--tpp_ada_temp] --seed N
```

### W&B runs

- GPU 0 (Gumbel+AdaTemp, s42): `udnzc9na` | GPU 1 (s43): `ttdgobca` | GPU 2 (s44): `5kyaycy5` | GPU 3 (s45): `n2o1lnz0`
- GPU 4 (Gumbel, s42): `609h12dx` | GPU 5 (s43): `7d1jhn5p`
- GPU 6 (Baseline NC, s42): `l8b0i0ny` | GPU 7 (s43): `70bmrhoc`

### What happened

The premise that compile takes 114 min of 180 min appears to have been from a specific configuration (possibly with Transolver-3 features that hit inductor bugs). For the standard Transolver with our baseline config, torch.compile:
1. Takes ~10-15 min to compile on the first epoch
2. Then runs ~2x faster per epoch for the remaining 165+ minutes
3. Net result: 160 epochs vs 68-74 without compile

**Disabling compile is always worse** for configurations that compile successfully. The right approach for features that break torch.compile is to fix the compilation issue rather than disable compile.

### Suggested follow-ups

1. **Test Gumbel+AdaTemp WITH torch.compile enabled** — the features themselves may help if given the full 160-epoch budget. Need to verify they compile successfully.
2. **If compile breaks with Gumbel**: The Gumbel noise (`torch.zeros_like().uniform_()`) should be compile-compatible. AdaTemp's mean-pool + linear should also compile fine. If issues arise, they might need `torch.compiler.disable` only for specific ops.
3. **Don't disable compile globally** — it's a 2x throughput regression.

**Comment by tcapelle** [comment]
ADVISOR: ## Review — Transolver++ Without torch.compile

**Result: CLOSE — Dead end, but valuable insight captured.**

Disabling torch.compile is catastrophic — 2x slower per epoch (148-160s vs ~75s), only 68-74 epochs vs ~160 in the same 180-min window.

**Within no-compile (apples-to-apples):**

| Config | Mean val/loss | Mean p_in |
|--------|--------------|-----------|
| Baseline (no-compile) | 0.530 | 22.64 |
| Gumbel+AdaTemp | 0.555 | 24.35 |
| Gumbel only | 0.561 | 23.82 |

The T++ features actually perform *slightly worse* than baseline even when controlling for throughput. This adds evidence to our prior finding that Gumbel-Softmax and per-point adaptive temperature are null results on our problem.

**Critical insight from this PR:** torch.compile takes only ~10-15 min for our standard Transolver config (not 114 min). The 114-min compile time was specific to Transolver-3's incompatible ops. This means **any modifications that compile cleanly get the full 2x speedup** — we should always keep compile enabled.

**Key takeaway for all future experiments:** Never disable torch.compile globally. If a feature breaks compile, fix the specific incompatible op or wrap it with `torch.compiler.disable`, don't disable the entire compiler.

Great characterization work, thorfinn. The compile timing data is especially valuable.
